internal/enricher/geolocation.go

```diff
@@ -0,0 +1,235 @@
+	Postal   string `json:"postal"`
+	Location string `json:"location"`
+	Timezone string `json:"timezone"`
+}
+
+type geoEnricher struct {
+	fetcher geolocation.GeoFetcher
+	logger  logger.Logger
+	stats   stats.Stats
+}
+
+func NewGeoEnricher(conf *config.Config, log logger.Logger, statClient stats.Stats) (PipelineEnricher, error) {
+	log.Infof("Setting up new geo enricher")
+
+	dbPath, err := downloadMaxmindDB(context.Background(), conf, log)
+	if err != nil {
+		return nil, fmt.Errorf("downloading instance of maxmind db: %w", err)
+	}
+
+	fetcher, err := geolocation.NewMaxmindDBReader(dbPath)
+	if err != nil {
+		return nil, fmt.Errorf("creating new instance of maxmind's geolocation db reader: %w", err)
+	}
+
+	return &geoEnricher{
+		fetcher: fetcher,
+		stats:   statClient,
+		logger:  log.Child("geolocation"),
+	}, nil
+}
+
+// Enrich function runs on a request of GatewayBatchRequest which contains
+// multiple singular events from a source. The enrich function augments the
+// geolocation information per event based on IP address.
+func (e *geoEnricher) Enrich(source *backendconfig.SourceT, request *types.GatewayBatchRequest) error {
+	if !source.GeoEnrichment.Enabled {
+		return nil
+	}
+	e.logger.Debugw("received a call to enrich gateway events for source", "sourceID", source.ID)
+
+	var (
+		rawGeo *geolocation.GeoInfo
+		err    error
+	)
+
+	defer func() {
+		errType := ""
+
+		if request.RequestIP == "" {
+			errType = ERR_EMPTY_IP
+		}
+
+		if err != nil {
+			errType = ERR_LOCATE_FAILED
+
+			if errors.Is(err, geolocation.ErrInvalidIP) {
+				errType = ERR_INVALID_IP
+			}
+		}
+
+		e.stats.NewTaggedStat(
+			"proc_geo_enricher_request",
+			stats.CountType,
+			stats.Tags{
+				"sourceID": source.ID,
+				"error":    errType,
+			}).Increment()
+	}()
+
+	if request.RequestIP == "" {
+		return nil
+	}
+
+	defer e.stats.NewTaggedStat(
+		"proc_geo_enricher_request_latency",
+		stats.TimerType,
+		stats.Tags{
+			"sourceID": source.ID,
+		},
+	).RecordDuration()()
+
+	rawGeo, err = e.fetcher.Locate(request.RequestIP)
+	if err != nil {
+		return fmt.Errorf("enriching request with geolocation: %w", err)
+	}
+
+	geoData := extractGeolocationData(rawGeo)
+
+	for _, event := range request.Batch {
+		// if the context section is missing on the event
+		// set it with default as map[string]interface{}
+		if _, ok := event["context"]; !ok {
+			event["context"] = map[string]interface{}{}
+		}
+
+		if context, ok := event["context"].(map[string]interface{}); ok {
+			// if the `geo` key already present fire a stat for observability
+			if _, ok := context["geo"]; ok {
+				continue
+			}
+			// update the geo section
+			context["geo"] = geoData
+		}
+	}
+
+	return nil
+}
+
+func (e *geoEnricher) Close() error {
+	if err := e.fetcher.Close(); err != nil {
+		return fmt.Errorf("closing the geo enricher: %w", err)
+	}
+	return nil
+}
+
+func downloadMaxmindDB(ctx context.Context, conf *config.Config, log logger.Logger) (string, error) {
```

Comment:
some doc would be nice, e.g. downloads the file from s3 if not already downloaded